libvast/vast/table_slice.hpp

```diff
@@ -28,6 +28,119 @@
 
 namespace vast {
 
+/// A horizontal partition of a table. A slice defines a tabular interface for
+/// accessing homogenous data independent of the concrete carrier format.
+class table_slice final {
+public:
+  // -- member types -----------------------------------------------------------
+
+  /// Platform-independent unsigned integer type used for sizes.
+  using size_type = uint64_t;
+
+  /// The possible encodings of a table slice.
+  /// @note This encoding is unversioned. Newly created table slices are
+  /// guaranteed to use the newest vesion of the encoding, while deserialized
+  /// table slices may use an older version.
+  enum class encoding : uint8_t {
+    none,  ///< No data is encoded; the table slice is empty or invalid.
+    arrow, ///< The table slice is encoded using the column-major Apache Arrow.
+    msgpack, ///< The table slice is encoded using the row-major MessagePack.
+  };
+
+  // -- constructors, destructors, and assignment operators --------------------
+
+  /// Default-constructs an empty table slice.
+  table_slice() noexcept;
+
+  /// Construct a table slice from a chunk of data.
+  /// @note Constructs an invalid table slice if the verification fails.
+  explicit table_slice(chunk_ptr&& chunk, bool verify = true) noexcept;
```

Comment:
It would be great if you added `enum class verify : bool { off, on };` for this. booleans parameters are annoying at the call site. I'm also thinking that it might be better to not have a default here.